File: encryption.py
import random
import math


def gcd(a, b):
    while b != 0:
        a, b = b, a % b
    return a


def isCoprime(e, lcm):
    return gcd(e, lcm) == 1


def findLCM(num1, num2):
    num1 -= 1
    num2 -= 1
    if num1 > num2:
        greater = num1
    else:
        greater = num2

    while (True):
        if (greater % num1 == 0) and (greater % num2 == 0):
            lcm = greater
            break
        greater += 1

    return lcm


def primesInRange(x, y):  # Method to calculate all primes in a given range from x to y
    prime_list = []  # list for storing said prime numbers
    for n in range(x, y):
        isPrime = True

        for num in range(2, n):
            if n % num == 0:
                isPrime = False

        if isPrime:
            prime_list.append(n)
    return prime_list


def modInv(u, v):
    for x in range(1, v):
        if ((u % v) * (x % v) % v == 1):
            return x


def getCoprime(lcm):
    for i in range(2, lcm):
        if isCoprime(i, lcm) == True:
            return i


def getTotient(x, y):
    return (x - 1) * (y - 1)


def getDval(phi, e):
    for i in range(e * 25):
        if (e * i) % phi == 1:
            return i


def getKeys():
    primes_list = primesInRange(200, 600)

    q = random.choice(primes_list)  # first prime number
    p = random.choice(primes_list)  # second prime number

    n = p * q  # n is the product of the two generated prime numbers
    print("p: ", p)
    print("p: ", q)
    print("n: ", n)

    phi = getTotient(p, q)
    print("phi: ", phi)
    lcm = findLCM(p, q)
    print("lcm: ", lcm)

    coprime = getCoprime(lcm)

    print("coprime:", coprime)
    phi = getTotient(p, q)
    d = modInv(coprime, phi)  # modinv(phi, coprime)
    print("d: ", d)

    publicKey = [n, coprime]
    privateKey = [n, d]

    return [publicKey, privateKey]


def printKeys(keys: list):
    keyFile = open('keys.txt', 'w')
    print(str(keys[0][0]) + ", " + str(keys[0][1]), file=keyFile)
    print(str(keys[1][0]) + ", " + str(keys[1][1]), file=keyFile)
    keyFile.close()


def encryptToFile(messageList: list , publicKey: list):
    encryptFile = open('message.txt', 'w')
    for i in range(len(messageList)):
        print(pow(ord(message[i]), publicKey[1]) % publicKey[0], file=encryptFile)


def encryptToTerminal(messageList: list , publicKey: list):
    for i in range(len(messageList)):
        print(pow(ord(message[i]), publicKey[1]) % publicKey[0])



message = 'Hello World'

keys = getKeys()
printKeys(keys)
publicKey = keys[0]
privateKey = keys[1]

print("Public key", keys[0])
print("Private key", keys[1])

print("Original Message: ", message)
encryptToFile(list(message), publicKey)
encryptToTerminal(list(message), publicKey)
#encryptedMessage = pow(message, publicKey[1]) % publicKey[0]
#print("Encrypted Message: ", encryptedMessage)


#decryptedMessage = pow(encryptedMessage, privateKey[1]) % privateKey[0]
#print("Decrypted Message IN ENCRYPT: ", decryptedMessage)

File: encryption.cpp
#include <iostream>
#include <math.h>
#include <ap_int.h>


using namespace std;
typedef ap_uint<1024> edmessage;

#define BITS 16

int getGDC(int a, int b);
int main();
int modinv(int a, int m);
int modexp(edmessage base, edmessage exp, int modulo);
bool isCoprime(int e, int lcm);
int getCoprime(int lcm);
int getTotient(int x, int y);
int findLCM(int a, int b);
bool isPrime(int n);
edmessage edPow(edmessage base, int toPow);


/**
 * @brief
 *
 * @param n
 * @return true
 * @return false
 */
bool isPrime(int n) {
    if(n == 0 || n == 1) {
        return false;
    }

    for(int i = 2; i <= n/2; i++) {
        if(n % i == 0) {
            return false;
        }

    }
    return true;
}


/**
 * @brief
 *
 * @param a
 * @param b
 * @return int
 */
int getGDC(int a, int b) {
    int c;
    while(b != 0) {
        c = a % b;
        a = b;
        b = c;
    }
    return a;
}

/**
 * @brief
 *
 * @param a
 * @param m
 * @return int
 */
int modinv(int a, int m) {
    int m0 = m;
    int y = 0;
    int x = 1;
    if(m == 1) {
        return 0;
    }

    while(a > 1) {
        int q = a / m;  //calculate the quotient
        int t = m;

        m = a % m, a = t;
        t = y;

        y = x - q * y;
        x = t;
    }
    if(x < 0) {
        x += m0;
    }
    return x;
}

/**
 * @brief
 *
 * @param base
 * @param exp
 * @param modulo
 * @return int
 */
int modexp(edmessage base, edmessage exp, int modulo) {
	edmessage sol = 1;
    while(exp > 0) {
        //cout << "base: " << base << endl;
        //cout << "exp: " << exp << endl;
        //cout << "modulo: " << modulo << endl;

        if(exp % 2 == 1) {
            //cout << "sol = " << sol << " * " << base << " % " << modulo << endl;
            sol = (sol * base) % modulo;
        }
        exp = exp >> 1;
        //cout << "base = " << base << " * " << base << " % " << modulo << endl;
        edmessage base = (base * base) % modulo;
        //cout << "new base: " << base << endl;
        //cout << "--------------------------------" << endl;

    }
    return sol;
}

/**
 * @brief
 *
 * @param e
 * @param lcm
 * @return true
 * @return false
 */
bool isCoprime(int e, int lcm) {
    if(isPrime(e)) {
        return getGDC(e, lcm) == 1;
    }
}

/**
 * @brief Calculate a coprime of the least common factor
 *
 * @param lcm Least common factor
 * @return A coprime of lcm
 */
int getCoprime(int lcm) {
    for(int i = lcm; i > 2; i--) {
        if(isCoprime(i, lcm) == true) {
            return i;
        }
    }
}

/**
 * @brief Calculates the least common multiple of a and b
 *
 * @param a First integer input
 * @param b Second integer input
 * @return The least common multiple of a and b
 */
int findLCM(int a, int b) {
    int greater;
    int lcm;

    a -= 1;
    b -= 1;

    if(a > b) {
        greater = a;
    } else {
        greater = b;
    }
    while (true)
    {
        if(greater % a == 0 && greater % b == 0) {
            lcm = greater;
            break;
        }
        greater++;
    }
    return lcm;
}


edmessage edPow(edmessage base, int toPow){
	edmessage count;
	count = 1;
	for(int i =0; i < toPow; i++){
		count *= base;
	}
	return count;
}


int test(int message, int &privateKey, int &publicKey, int &n) {
    int p = 11;
    int q = 17;

    cout << "p: " << p << " q: " << q << endl;

     n = p * q;
    cout << "n: " << n << endl;

    int lcm = findLCM(p, q);
    cout << "lcm: " << lcm << endl;

    int coprime = 23; //getCoprime(lcm);
    cout << "coprime: " << coprime << endl;

    int d = modinv(coprime, lcm);
    cout << "d: " << d << endl;
    privateKey = d;
    publicKey = coprime;
    cout << "Original message: " << message << endl;


    cout << "String message converted to ints: ";


    cout << endl;



     cout << "Message: " << message << endl;
     // edmessage encryptedMessage = fmod((edmessage)pow(message, coprime), n);
    int  encryptedMessage = edPow((edmessage)message, coprime) % n;
     cout << "Encryptd Message: " << encryptedMessage << endl;


    //  edmessage decryptedMessage = fmod((edmessage)pow(encryptedMessage, d),n);
      edmessage decryptedMessage = edPow(encryptedMessage, d) % n;
      cout << "Decrypted Message: " << decryptedMessage << endl;

      return encryptedMessage;
}
